### ISSUE TITLE
App sharing does not work

### ISSUE THREAD

**Issue by Demsli** [ronaldoussoren/py2app#492]
Greetings everyone. I am trying to freeze my program and share it with my colleagues. When I freeze the program on my MacBook Air M2 - the program works. But as soon as I compress it into the zip folder and share it (for example via telegram or google disk) I get a message "The application can't be opened". When going into the contents of the app, into macOS folder and launching the unix file I get an error that the file is damaged and should be deleted. I am doing this testing of download and unpackaging on the same Air, which is even more weird why this occurs, since it is the one that build the app. Can it be the problem of app signing?
**Py2app version:** 0.28.5
**Python version:** 3.11.3
**macOS version:** 12.5
**setup code:**
from setuptools import setup

APP = ['diviner.py']
DATA_FILES = []
OPTIONS = {'iconfile': '/Users/profix/PS-mac.icns'}

setup(
    app=APP,
    data_files=DATA_FILES,
    options={'py2app': OPTIONS},
    setup_requires=['py2app'],
)
**Program code:**
import tkinter as tk

def divide_by_two():
    try:
        # Get the number from the entry widget
        number = float(entry.get())
        
        # Divide the number by two
        result = number / 2
        
        # Update the label with the result
        result_label.config(text=f"Result: {result}")
    except ValueError:
        # Display an error message if the input is not a valid number
        result_label.config(text="Invalid input")

# Create the Tkinter window
window = tk.Tk()
window.title("Number Divider")
window.geometry("200x200")

# Create the Entry widget
entry = tk.Entry(window)
entry.pack()

# Create the button
button = tk.Button(window, text="Divide by 2", command=divide_by_two)
button.pack()

# Create the label
result_label = tk.Label(window, text="Result: ")
result_label.pack()

# Start the Tkinter event loop
window.mainloop()
The program is simple, I am just using it as a test to see how freezing on Mac works. The app freezes, but what's the point if I cannot share it with my friends?)

**Comment by ronaldoussoren** [comment]
@StefanJordan mention in #493 that removing extended attributes can help (``xattr -cr /path_to_App/App.app`` in the terminal). 

That points to the macOS quarantine system interfering. I hope to have time over the summer to investigate this fully and work on a better solution.

**Comment by Demsli** [comment]
> Does this error also occur when you open the zipfile locally and launch the application?

Nope, compressing and decompressing immediately works. Downloading the zip from external source causes the issue of "Not being able to open"

**Comment by ronaldoussoren** [comment]
Either remove extended attributes from the app bundle (see my previous comment), or try opening using the context menu (right clicking) instead of double clicking. The latter tends to be more forgiving in what it accepts, especially when app bundles are not signed using a developer ID.

**Comment by Demsli** [comment]
> Either remove extended attributes from the app bundle (see my previous comment), or try opening using the context menu (right clicking) instead of double clicking. The latter tends to be more forgiving in what it accepts, especially when app bundles are not signed using a developer ID.

xattr -cr gave me an error "option -r not recognized", so I used xattr -c. Still same error(

**Comment by StefanJordan** [comment]
Demsli, did you pack it into a dmg image first? It did not work without doing it because the file got corrupted whenever a zip was involved in the distribution! Maybe tar would work but I am not sure. 

And do not apply it to the dmg file, of course, but to the app. Maybe it does not always work with this method but I was successful with it.

**Comment by Demsli** [comment]
I applied it to the file and tried to zip it. Tried the sudo method as well - now after unpacking from download it says "file not found" though the app is there. Weird

**Comment by Demsli** [comment]
Will dmg make a difference?

**Comment by StefanJordan** [comment]
Yes, it made the difference between it works and it does not work. 
 Drag the folder onto the Disk Utility-icon, or choose "New->Disk image" from folder in the File-menu in Disk Utility.

**Comment by Demsli** [comment]
"Application diviner can't be opened" :(

**Comment by StefanJordan** [comment]
Did you extract the app from the dmg file AND applied "sudo xattr -cr /path/App.app"?

**Comment by Demsli** [comment]
My bad, thought that I had to apply first, dmg it and then send. Will try this order

**Comment by Demsli** [comment]
1. Ran python3 setup.py py2app
2. Disk Utility -> New image from folder -> .app
3. Sent diviner.dmg to telegram and downloaded back to the machine
4. opened diviner.dmg and got the file out
5. In terminal located the folder
6. Applied sudo xattr -c diviner.app
7. Still got the error that diviner can't be opened(

**Comment by StefanJordan** [comment]
create app -> packing app into dmg -> deliver -> unpack app -> move to its destination ->  apply "sudo xattr -cr /path/App.app"

**Comment by Demsli** [comment]
How can I check that xattr applied?

**Comment by ronaldoussoren** [comment]
> Did you extract the app from the dmg file AND applied "sudo xattr -cr /path/App.app"?

The ``sudo`` part should not be necessary if the app bundle is owned by yourself (that is, not a system binary). 

Also for anyone not familiar with the command-line, please use a tool like "standstrip" to remove unwanted extended attributes: https://eclecticlight.co/xattred-sandstrip-xattr-tools/.

**Comment by StefanJordan** [comment]
I have no idea. Try opening it with the right mouse click. I only can tell what worked for me with my app and three others who are using it.

**Comment by Demsli** [comment]
> I have no idea. Try opening it with the right mouse click. I only can tell what worked for me with my app and three others who are using it.

do you have the arm based MacBook or intel one? can the architecture be the case?

**Comment by StefanJordan** [comment]
Mac Book Pro with M1 in my case. And Ventura 13.4 in my case and in the case of the ones who use the app. I am not sure whether they have Intel or M1/M2.

**Comment by Demsli** [comment]
> > Did you extract the app from the dmg file AND applied "sudo xattr -cr /path/App.app"?
> 
> The `sudo` part should not be necessary if the app bundle is owned by yourself (that is, not a system binary).
> 
> Also for anyone not familiar with the command-line, please use a tool like "standstrip" to remove unwanted extended attributes: https://eclecticlight.co/xattred-sandstrip-xattr-tools/.

checked the file after extraction with xattred. It did have com.apple.quarantine tag with string 0187;6488c803;Telegram; attached. Tried to cut it - it created com.apple.provenance tag, but did not solve the problem, the app still does not open:(

**Comment by Demsli** [comment]
Tried sand strip, got a new error - “diviner” can’t be opened because Apple cannot check it for malicious software.

**Comment by StefanJordan** [comment]
Maybe you could follow https://iboysoft.com/news/app-is-damaged-and-cannot-be-opened.html#how-to-fix-fileapp-is-damaged-and-cannot-be-opened-on-macos-venturabig-sur
If you are experiencing "App is damaged and cannot be opened" on macOS Ventura, do the following steps to fix it:

Open the Apple menu > [System Settings](https://iboysoft.com/wiki/macos-ventura-system-settings.html).
Select Privacy & Security > Developer Tools.
Click the ( + ) button and navigate to the folder where the damaged app resides. 
Select the app and click Open.
If you are encountering "Application is damaged and cannot be opened" on Big Sur/Monterey/Catalina when opening an app, try these steps:

Open the Apple menu > System Preferences.
Select Security & Privacy.
Tap the yellow lock and enter your password to unlock the preference pane.
Click "Open Anyway."


Maybe this help. I have no idea...

**Comment by Demsli** [comment]
I will try to use the zip method. I will also see if stripping the archive works

**Comment by Demsli** [comment]
Stripping the archive itself and unzipping also works. Will try to upload the stripped archive again. To see if it will have the quarantine tag the second time

**Comment by Demsli** [comment]
Yes, sending the stripped archive adds the tag again. So, in order to share the program, the steps are:
1. Freeze the code
2. Compress the app into .zip archive
3. Send the .zip to a friend
4. Friend unzips the file
5. Uses Sandstrip to strip the quarantine tag
6. Opens the app to receive the error of unidentified app
7. Right clicks the app and selects open
8. Now the second option "Open" will appear
9. Clicking "Open" in the window will open the app. Now the system trust the app and it can be opened at any time.

**Comment by Demsli** [comment]
Will probably have to bundle the .zip archive with the program and the necessary steps. Will also test the original program I needed to freeze. Thank you StefanJordan and ronaldoussoren for the help. Struggled with this thing for a while but glad that a solution was found)

**Comment by StefanJordan** [comment]
Yes, the problem is how to distribute it to people who are not very much into these technical steps. It would be good if one could bundle it in a way that the necessary steps are done automatically. Would a tool create a pkg file help in which a user could be just guided through the steps? I do not know such tools. I tried some packers but they did not work on Ventura.

**Comment by Demsli** [comment]
Sandstrip is really easy to use. you just select file and it automatically deletes all quarantine tags. But yeah, wish these steps were unnecessary. Yesterday I updated to Ventura and downloaded Xcode. I read that it is possible to sign the app with just  Apple ID. But not sure if it will work, the quarantine tag had Telegram in it. I bet that if I sent it to google disk the quarantine will remain, it probably gets added to anything that is not supported by Apple. Maybe dropbox will work? But for now the combo of stripper + app + some opening steps works.

**Comment by ronaldoussoren** [comment]
Standstrip is just a workaround. I will be looking into a better solution for this, but that will take time.  A future version of py2app will also have direct support for code signing.

**Comment by Demsli** [comment]
> May I ask if this issue has been resolved? I also encountered this problem. I tried to compress an. app file and share it with my friend, but his computer couldn't open it. My own computer is also unable to open compressed files. How can I solve this problem

As I stated a few messages above - I added Sandstrip to remove the quarantine tags from the file, then right clicked to open it. The popup will ask if you are sure to open this app. Click yes and the app should open. After that app will open as usual. I havent tried it on new versions of MacOS though

**Comment by hb-2017** [comment]
Yes, I tried using Sandstrip, but I didn't know how to use it after downloading. Can you briefly explain how to use it? I don't even know if the version I downloaded is correct

**Comment by hb-2017** [comment]
I think I have found the correct version, but even after using the compressed and decompressed. app file, I still cannot open it

**Comment by Demsli** [comment]
> I think I have found the correct version, but even after using the compressed and decompressed. app file, I still cannot open it

What MacOS version do you have? And what's the version of py2app you use? And just in case - what's you mac model (M chip or Intel)?

**Comment by hb-2017** [comment]
I would like to confirm the process here.
1. First, download the Sandstrip tool.
![image](https://github.com/user-attachments/assets/2ef1f942-c8f3-4477-99b4-5bd9e4537279)

<img width="305" alt="image" src="https://github.com/user-attachments/assets/f6cfae05-1b12-43c4-8c46-b095be856faa">

2. Then use the tool to select the decompressed. app file
<img width="883" alt="image" src="https://github.com/user-attachments/assets/304a31e9-1fbe-4773-8d8b-ce364857ebc7">

3. Right click on the standalone device and open it


After doing this, I still can't open it

**Comment by Demsli** [comment]
Hmm. That's a lot of files stripped. You can also try xattred to see if it will work instead. I see that it's actually xattred that shows the download source, not Sandstrip, my bad.

**Comment by hb-2017** [comment]
<img width="480" alt="image" src="https://github.com/user-attachments/assets/e49cc05e-6aaf-48c2-b1c6-86430f162335">

After checking with Xattred, I found that it only has 2 flags. But I still can't open him

**Comment by hb-2017** [comment]
Is there any other way to share with friends? He can accept any form of file as long as it is runnable

**Comment by Demsli** [comment]
Found a video that I made for my app. Used Xattred there. You need to choose the quarantine tag and clip strip flag:
![image](https://github.com/user-attachments/assets/a22210a9-a0cd-4201-8456-84879c53371f)

**Comment by Demsli** [comment]
Oh shoot, you dont have the quarantine flag

**Comment by hb-2017** [comment]
Should I not use Sandstrip? But instead use Xattred directly? If I were to use xattred directly, how should I choose the flag that I need to strip

**Comment by Demsli** [comment]
> Should I not use Sandstrip? But instead use Xattred directly? If I were to use xattred directly, how should I choose the flag that I need to strip

Yeah, try to decompress the app again and use Xattred this time. There is a chance that Sandstrip deleted too many files that the app is now not working at all

**Comment by hb-2017** [comment]
I found that I have already solved this problem. I just tried using the built-in compression tool on my Mac computer, and after decompressing it, it can run normally. After sharing it with friends, it can also run normally. Before this, I used a third-party compression tool, which should be the problem.

Although I didn't understand the principle here, my problem has been solved. Thank you very much
